File: internal/core/agent/loop.go
package agent

import (
	"context"
	"encoding/json"
	"fmt"
	"log/slog"

	"github.com/collinpfeifer/beluga/internal/core/eventstore"
	"github.com/collinpfeifer/beluga/internal/core/model"
	"github.com/collinpfeifer/beluga/internal/core/session"
)

// ToolExecutor is the interface for executing tools in the agent loop.
// This decouples the orchestrator from the tools package.
type ToolExecutor interface {
	ExecuteTool(ctx context.Context, name string, args json.RawMessage, sessionID string) (json.RawMessage, error)
}

// Orchestrator manages agent session lifecycles and runs the agent loop.
type Orchestrator struct {
	sessions     *session.Store
	events       *eventstore.Store
	llm          *LLMClient
	ctxBuilder   *ContextBuilder
	compactor    *Compactor
	toolExecutor ToolExecutor
	logger       *slog.Logger
}

// New creates a new orchestrator.
func New(
	sessions *session.Store,
	events *eventstore.Store,
	llm *LLMClient,
	toolExecutor ToolExecutor,
	logger *slog.Logger,
) *Orchestrator {
	ctxBuilder := NewContextBuilder()
	return &Orchestrator{
		sessions:     sessions,
		events:       events,
		llm:          llm,
		toolExecutor: toolExecutor,
		ctxBuilder:   ctxBuilder,
		compactor:    NewCompactor(events, llm, ctxBuilder, logger),
		logger:       logger,
	}
}

// NewWithPrompt creates a new orchestrator with a custom system prompt.
func NewWithPrompt(
	sessions *session.Store,
	events *eventstore.Store,
	llm *LLMClient,
	toolExecutor ToolExecutor,
	systemPrompt string,
	logger *slog.Logger,
) *Orchestrator {
	ctxBuilder := NewContextBuilderWithPrompt(systemPrompt)
	return &Orchestrator{
		sessions:     sessions,
		events:       events,
		llm:          llm,
		toolExecutor: toolExecutor,
		ctxBuilder:   ctxBuilder,
		compactor:    NewCompactor(events, llm, ctxBuilder, logger),
		logger:       logger,
	}
}

// SetTools wires tool definitions into the context builder so they are
// included in LLM chat completion requests.
func (o *Orchestrator) SetTools(tools []ToolDef) {
	o.ctxBuilder.SetTools(tools)
}

// StartRun begins an agent loop for the given session.
func (o *Orchestrator) StartRun(ctx context.Context, sessionID string) error {
	sess, err := o.sessions.Get(ctx, sessionID)
	if err != nil {
		return fmt.Errorf("getting session: %w", err)
	}

	if sess.Status != model.StatusPending && sess.Status != model.StatusSuspended {
		return fmt.Errorf("cannot start run for session in status %s", sess.Status)
	}

	// Transition to running.
	if err := o.sessions.UpdateStatus(ctx, sessionID, model.StatusRunning); err != nil {
		return fmt.Errorf("updating status to running: %w", err)
	}

	// Record the status transition event.
	if _, err := o.appendTransition(ctx, sessionID, sess.Status, model.StatusRunning); err != nil {
		return fmt.Errorf("recording status transition: %w", err)
	}

	// Run the agent loop in a goroutine.
	go o.runLoop(context.Background(), sessionID)

	return nil
}

// Interrupt gracefully interrupts a running session.
func (o *Orchestrator) Interrupt(ctx context.Context, sessionID string) error {
	sess, err := o.sessions.Get(ctx, sessionID)
	if err != nil {
		return fmt.Errorf("getting session: %w", err)
	}

	if sess.Status != model.StatusRunning {
		return fmt.Errorf("cannot interrupt session in status %s", sess.Status)
	}

	if err := o.sessions.UpdateStatus(ctx, sessionID, model.StatusSuspended); err != nil {
		return fmt.Errorf("updating status to suspended: %w", err)
	}

	_, _ = o.events.Append(ctx, sessionID, model.EventTypeInterrupt, mustJSON(model.InterruptPayload{
		Reason:         "interrupted by user",
		RequiresAction: true,
	}))

	_, _ = o.appendTransition(ctx, sessionID, model.StatusRunning, model.StatusSuspended)
	return nil
}

// Resume resumes a suspended session.
func (o *Orchestrator) Resume(ctx context.Context, sessionID string) error {
	return o.StartRun(ctx, sessionID)
}

// Terminate kills the session and marks it completed or failed.
func (o *Orchestrator) Terminate(ctx context.Context, sessionID string, failed bool) error {
	sess, err := o.sessions.Get(ctx, sessionID)
	if err != nil {
		return fmt.Errorf("getting session: %w", err)
	}

	newStatus := model.StatusCompleted
	if failed {
		newStatus = model.StatusFailed
	}

	if err := o.sessions.UpdateStatus(ctx, sessionID, newStatus); err != nil {
		return fmt.Errorf("updating status to %s: %w", newStatus, err)
	}

	_, _ = o.appendTransition(ctx, sessionID, sess.Status, newStatus)

	o.logger.Info("session terminated",
		"session_id", sessionID,
		"status", newStatus,
	)
	return nil
}

// HandleNewSession creates a new session and starts the agent loop.
// This is the callback passed to extensions via ExtensionContext.CreateSession.
func (o *Orchestrator) HandleNewSession(ctx context.Context, source, sourceID string, initialMessage string, metadata json.RawMessage) (*model.Session, error) {
	sess, err := o.sessions.Create(ctx, source, sourceID, metadata)
	if err != nil {
		return nil, fmt.Errorf("creating session: %w", err)
	}

	// Seed the session with the initial user message if provided.
	if initialMessage != "" {
		if _, err := o.events.Append(ctx, sess.ID, model.EventTypeUserMessage, mustJSON(model.UserMessagePayload{
			Content: initialMessage,
		})); err != nil {
			return nil, fmt.Errorf("seeding user message: %w", err)
		}
	}

	if err := o.StartRun(ctx, sess.ID); err != nil {
		return nil, fmt.Errorf("starting run: %w", err)
	}

	return sess, nil
}

// runLoop is the main agent loop. It runs in a goroutine.
func (o *Orchestrator) runLoop(ctx context.Context, sessionID string) {
	logger := o.logger.With("session_id", sessionID)
	logger.Info("agent loop started")

	const maxIterations = 30
	for iteration := 0; iteration < maxIterations; iteration++ {
		// Check if session is still running.
		sess, err := o.sessions.Get(ctx, sessionID)
		if err != nil {
			logger.Error("failed to get session", "error", err)
			_ = o.Terminate(ctx, sessionID, true)
			return
		}
		if sess.Status != model.StatusRunning {
			logger.Info("session no longer running, stopping loop", "status", sess.Status)
			return
		}

		// Build context from session events.
		events, err := o.events.GetEvents(ctx, sessionID, 0, 10000)
		if err != nil {
			logger.Error("failed to get events", "error", err)
			_ = o.Terminate(ctx, sessionID, true)
			return
		}

		messages := o.ctxBuilder.Build(events)

		// Check if context needs compaction before calling the LLM.
		if o.compactor.NeedsCompaction(messages) {
			logger.Info("context approaching limit, triggering compaction",
				"estimated_tokens", o.ctxBuilder.EstimateTokens(messages),
				"max_tokens", o.ctxBuilder.MaxTokens(),
			)
			tokensSaved, err := o.compactor.Compact(ctx, sessionID, events)
			if err != nil {
				logger.Error("compaction failed", "error", err)
				// Don't terminate — just continue with the truncated context.
			} else if tokensSaved > 0 {
				// Re-fetch events and rebuild context after compaction.
				events, err = o.events.GetEvents(ctx, sessionID, 0, 10000)
				if err != nil {
					logger.Error("failed to get events after compaction", "error", err)
					_ = o.Terminate(ctx, sessionID, true)
					return
				}
				messages = o.ctxBuilder.Build(events)
				logger.Info("context compacted, rebuilt messages",
					"estimated_tokens", o.ctxBuilder.EstimateTokens(messages),
					"tokens_saved", tokensSaved,
				)
			}
		}

		logger.Info("calling LLM",
			"model", o.llm.Model(),
			"message_count", len(messages),
			"tools_count", len(o.ctxBuilder.tools),
		)

		resp, err := o.llm.Chat(ctx, ChatRequest{
			Model:    "", // use default from config
			Messages: messages,
			Tools:    o.ctxBuilder.tools,
		})
		if err != nil {
			logger.Error("LLM call failed", "error", err)
			_, _ = o.events.Append(ctx, sessionID, model.EventTypeError, mustJSON(model.ErrorPayload{
				Message:     fmt.Sprintf("LLM call failed: %v", err),
				Recoverable: true,
			}))
			_ = o.Terminate(ctx, sessionID, true)
			return
		}

		pt, ct, tt := 0, 0, 0
		if resp.Usage != nil {
			pt = resp.Usage.PromptTokens
			ct = resp.Usage.CompletionTokens
			tt = resp.Usage.TotalTokens
		}
		logger.Info("LLM response received",
			"choices", len(resp.Choices),
			"prompt_tokens", pt,
			"completion_tokens", ct,
			"total_tokens", tt,
		)

		if len(resp.Choices) == 0 {
			logger.Error("LLM returned no choices")
			_ = o.Terminate(ctx, sessionID, true)
			return
		}

		choice := resp.Choices[0]

		logger.Info("LLM choice",
			"finish_reason", choice.FinishReason,
			"content_len", len(choice.Message.Content),
			"tool_calls", len(choice.Message.ToolCalls),
		)

		// Handle tool calls.
		if len(choice.Message.ToolCalls) > 0 {
			for _, tc := range choice.Message.ToolCalls {
				// Record the tool call event.
				_, _ = o.events.Append(ctx, sessionID, model.EventTypeToolCall, mustJSON(model.ToolCallPayload{
					ToolName: tc.Function.Name,
					Args:     json.RawMessage(tc.Function.Arguments),
					CallID:   tc.ID,
				}))

				// Execute the tool via the tool executor.
				var result json.RawMessage
				var toolErr error
				if o.toolExecutor != nil {
					result, toolErr = o.toolExecutor.ExecuteTool(ctx, tc.Function.Name, json.RawMessage(tc.Function.Arguments), sessionID)
				} else {
					result = nil
					toolErr = fmt.Errorf("tool %q execution not available", tc.Function.Name)
				}

				output := string(result)
				isError := false
				if toolErr != nil {
					output = toolErr.Error()
					isError = true
				}

				_, _ = o.events.Append(ctx, sessionID, model.EventTypeToolResult, mustJSON(model.ToolResultPayload{
					CallID:  tc.ID,
					Output:  output,
					IsError: isError,
				}))
			}
			// Continue the loop to process the next LLM turn.
			continue
		}

		// Handle agent message.
		if choice.Message.Content != "" {
			_, _ = o.events.Append(ctx, sessionID, model.EventTypeAgentMessage, mustJSON(model.AgentMessagePayload{
				Content:    choice.Message.Content,
				Model:      o.llm.Model(),
				TokensUsed: usageTokens(resp.Usage),
			}))

			// Check finish reason.
			switch choice.FinishReason {
			case "stop":
				// Agent is done.
				_ = o.Terminate(ctx, sessionID, false)
				return
			case "length":
				// Hit max tokens — continue the loop.
				logger.Info("LLM hit max tokens, continuing")
				continue
			default:
				logger.Info("LLM finished", "reason", choice.FinishReason)
				_ = o.Terminate(ctx, sessionID, false)
				return
			}
		}

		// No content and no tool calls — stop.
		logger.Warn("LLM returned empty message with no tool calls",
			"finish_reason", choice.FinishReason,
		)
		_ = o.Terminate(ctx, sessionID, false)
		return
	}
}

// appendTransition records a status transition event.
func (o *Orchestrator) appendTransition(ctx context.Context, sessionID, from, to string) (*model.Event, error) {
	return o.events.Append(ctx, sessionID, model.EventTypeStatusTransition, mustJSON(model.StatusTransitionPayload{
		From: from,
		To:   to,
	}))
}

// mustJSON marshals v to JSON, returning an empty object on error.
func mustJSON(v interface{}) json.RawMessage {
	data, err := json.Marshal(v)
	if err != nil {
		return json.RawMessage(`{}`)
	}
	return json.RawMessage(data)
}

// usageTokens extracts total tokens from a ChatUsage, returning 0 if nil.
func usageTokens(u *ChatUsage) int {
	if u == nil {
		return 0
	}
	return u.TotalTokens
}

File: src/core/agent/loop.ts
// ── Agent orchestrator ────────────────────────────────────────
// Ported from internal/core/agent/loop.go

import type { SessionStore } from "../session/store.js";
import type { EventStore } from "../eventstore/store.js";
import type { LLMClient } from "./llm.js";
import type { ContextBuilder } from "./context.js";
import type { Compactor } from "./compactor.js";
import type { ToolDef } from "../tools/registry.js";
import type { Session } from "../model/types.js";
import type { Logger } from "pino";

export interface ToolExecutor {
	executeTool(
		name: string,
		args: Record<string, unknown>,
		sessionId: string,
	): Promise<Record<string, unknown>>;
	/** Name of the agent this executor belongs to. */
	agent: string;
}

export class Orchestrator {
	private sessions: SessionStore;
	private events: EventStore;
	private llm: LLMClient;
	private toolExecutor: ToolExecutor;
	private contextBuilder: ContextBuilder;
	private compactor: Compactor;
	private logger: Logger;
	private tools: ToolDef[] = [];
	private maxIterations: number;
	private abortControllers = new Map<string, AbortController>();

	constructor(
		sessions: SessionStore,
		events: EventStore,
		llm: LLMClient,
		toolExecutor: ToolExecutor,
		contextBuilder: ContextBuilder,
		compactor: Compactor,
		maxIterations: number,
		logger: Logger,
	) {
		this.sessions = sessions;
		this.events = events;
		this.llm = llm;
		this.toolExecutor = toolExecutor;
		this.contextBuilder = contextBuilder;
		this.compactor = compactor;
		this.maxIterations = maxIterations;
		this.logger = logger;
	}

	setTools(tools: ToolDef[]): void {
		this.tools = tools;
		this.contextBuilder.setTools(tools);
	}

	/** Handle a new session: create it, seed initial message, start agent loop. */
	async handleNewSession(
		source: string,
		sourceId: string,
		initialMessage: string,
		metadata: Record<string, unknown> = {},
	): Promise<Session> {
		const session = await this.sessions.create(
			source,
			sourceId,
			metadata,
			this.toolExecutor.agent,
		);

		// Seed initial user message
		await this.events.append(session.id, "user_message", {
			content: initialMessage,
		});

		// Start agent loop
		this.startRun(session.id).catch(async (err) => {
			this.logger.error({ err, sessionId: session.id }, "agent loop failed");
			await this.terminate(session.id, true);
		});

		return session;
	}

	/** Continue an existing session with a new user message. Resumes if needed. */
	async handleContinueSession(
		sessionId: string,
		message: string,
		metadata?: Record<string, unknown>,
	): Promise<Session> {
		const session = await this.sessions.get(sessionId);
		if (!session) throw new Error(`session not found: ${sessionId}`);

		// Update metadata if provided
		if (metadata) {
			await this.sessions.updateMetadata(sessionId, metadata);
		}

		// Append new user message
		await this.events.append(sessionId, "user_message", {
			content: message,
		});

		// Resume if session was completed/failed — reset to pending first
		if (session.status === "completed" || session.status === "failed") {
			await this.sessions.updateStatus(sessionId, "pending");
			await this.events.append(sessionId, "status_transition", {
				from: session.status,
				to: "pending",
			});
			this.startRun(sessionId).catch(async (err) => {
				this.logger.error({ err, sessionId }, "agent loop failed on continue");
				await this.terminate(sessionId, true);
			});
		} else if (session.status === "running") {
			// Already running — the new message will be picked up on next iteration
			this.logger.info({ sessionId }, "session running, message queued");
		} else {
			// pending or suspended — start/resume
			this.startRun(sessionId).catch(async (err) => {
				this.logger.error({ err, sessionId }, "agent loop failed on continue");
				await this.terminate(sessionId, true);
			});
		}

		return session;
	}

	/** Start the agent loop for a session. */
	async startRun(sessionId: string): Promise<void> {
		const session = await this.sessions.get(sessionId);
		if (!session) throw new Error(`session not found: ${sessionId}`);

		if (session.status !== "pending" && session.status !== "running") {
			throw new Error(
				`session ${sessionId} is ${session.status}, cannot start`,
			);
		}

		await this.sessions.updateStatus(sessionId, "running");
		await this.events.append(sessionId, "status_transition", {
			from: session.status,
			to: "running",
		});

		// Register abort controller for this session
		const ac = new AbortController();
		this.abortControllers.set(sessionId, ac);

		this.logger.info({ sessionId }, "agent loop started");

		try {
			const maxIter = this.maxIterations || Infinity;
			for (let i = 0; i < maxIter; i++) {
				if (ac.signal.aborted) {
					this.logger.info({ sessionId, iteration: i }, "agent loop aborted");
					break;
				}
				const result = await this.runIteration(sessionId, ac.signal);
				if (result === "done") break;
			}
		} finally {
			this.abortControllers.delete(sessionId);
		}
	}

	private async runIteration(
		sessionId: string,
		signal: AbortSignal,
	): Promise<"continue" | "done" | "aborted"> {
		if (signal.aborted) return "aborted";

		// Fetch all events
		const allEvents = await this.events.getEvents(sessionId, 0, 10000);

		// Build context
		const messages = this.contextBuilder.build(allEvents);

		// Check if compaction needed
		if (this.compactor.needsCompaction(messages)) {
			await this.compactor.compact(sessionId, allEvents);
		}

		if (signal.aborted) return "aborted";

		// Build LLM tool definitions
		const { toLLMTools } = await import("../tools/registry.js");
		const llmTools = this.tools.length > 0 ? toLLMTools(this.tools) : undefined;

		// Call LLM
		const response = await this.llm.chat({
			messages,
			tools: llmTools,
		});

		if (signal.aborted) return "aborted";

		const choice = response.choices[0];
		if (!choice) {
			await this.terminate(sessionId, true);
			return "done";
		}

		// Handle tool calls
		if (choice.message.tool_calls && choice.message.tool_calls.length > 0) {
			for (const tc of choice.message.tool_calls) {
				if (signal.aborted) return "aborted";

				// Record tool_call event
				const args = JSON.parse(tc.function.arguments);
				this.logger.debug(
					{ sessionId, tool: tc.function.name, args },
					"tool call",
				);
				await this.events.append(sessionId, "tool_call", {
					tool_name: tc.function.name,
					args,
					call_id: tc.id,
				});

				try {
					// Execute tool
					const result = await this.toolExecutor.executeTool(
						tc.function.name,
						args,
						sessionId,
					);
					this.logger.debug(
						{ sessionId, tool: tc.function.name, result },
						"tool result",
					);

					await this.events.append(sessionId, "tool_result", {
						call_id: tc.id,
						output: JSON.stringify(result),
						is_error: false,
					});
				} catch (err) {
					this.logger.error(
						{ sessionId, tool: tc.function.name, err },
						"tool error",
					);
					await this.events.append(sessionId, "tool_result", {
						call_id: tc.id,
						output: String(err instanceof Error ? err.message : err),
						is_error: true,
					});
				}
			}
			return "continue";
		}

		// Handle content response — ALWAYS emit agent_message
		if (choice.message.content) {
			await this.events.append(sessionId, "agent_message", {
				content: choice.message.content,
				model: this.llm.model,
				usage: this.llm.lastUsage,
			});
		} else {
			// LLM returned empty content with no tool calls — send fallback so user gets a response
			this.logger.warn(
				{ sessionId, finishReason: choice.finish_reason },
				"LLM returned empty content, sending fallback message",
			);
			await this.events.append(sessionId, "agent_message", {
				content:
					"I encountered an issue processing your request. Please try again.",
				model: this.llm.model,
				usage: this.llm.lastUsage,
			});
		}

		// Check finish reason
		if (choice.finish_reason === "stop") {
			await this.terminate(sessionId, false);
			return "done";
		}

		if (choice.finish_reason === "length") {
			// Context window full — continue (compaction will handle it next iteration)
			return "continue";
		}

		// Any other finish reason — complete
		await this.terminate(sessionId, false);
		return "done";
	}

	/** Cancel a running agent loop. */
	async cancel(sessionId: string): Promise<boolean> {
		const ac = this.abortControllers.get(sessionId);
		if (!ac) return false;

		ac.abort();
		this.abortControllers.delete(sessionId);

		await this.sessions.updateStatus(sessionId, "completed");
		await this.events.append(sessionId, "status_transition", {
			from: "running",
			to: "completed",
			reason: "cancelled by user",
		});

		this.logger.info({ sessionId }, "agent loop cancelled by user");
		return true;
	}

	async interrupt(sessionId: string): Promise<void> {
		const ac = this.abortControllers.get(sessionId);
		if (ac) ac.abort();

		await this.sessions.updateStatus(sessionId, "suspended");
		await this.events.append(sessionId, "interrupt", {
			reason: "user requested interrupt",
			requires_action: false,
		});
	}

	async resume(sessionId: string): Promise<void> {
		await this.sessions.updateStatus(sessionId, "running");
		this.startRun(sessionId).catch((err) => {
			this.logger.error({ err, sessionId }, "agent loop failed on resume");
		});
	}

	async terminate(sessionId: string, failed: boolean): Promise<void> {
		const status = failed ? "failed" : "completed";
		await this.sessions.updateStatus(sessionId, status);
		await this.events.append(sessionId, "status_transition", {
			from: "running",
			to: status,
		});

		this.logger.info({ sessionId, status }, "session terminated");
	}
}
